codegen/src/error.rs

```diff
@@ -0,0 +1,120 @@
+    InvalidFields(String),
+    /// Substitute types must have a valid path.
+    #[error("Type substitution error: {0}")]
+    TypeSubstitutionError(#[from] TypeSubstitutionError),
+    /// Invalid type path.
+    #[error("Invalid type path {0}: {1}")]
+    InvalidTypePath(String, syn::Error),
+    /// Metadata for constant could not be found.
+    #[error("Metadata for constant entry {0}_{1} could not be found. Make sure you are providing a valid metadata V14")]
+    MissingConstantMetadata(String, String),
+    /// Metadata for storage could not be found.
+    #[error("Metadata for storage entry {0}_{1} could not be found. Make sure you are providing a valid metadata V14")]
+    MissingStorageMetadata(String, String),
+    /// Metadata for call could not be found.
+    #[error("Metadata for call entry {0}_{1} could not be found. Make sure you are providing a valid metadata V14")]
+    MissingCallMetadata(String, String),
+    /// Call variant must have all named fields.
+    #[error("Call variant for type {0} must have all named fields. Make sure you are providing a valid metadata V14")]
+    InvalidCallVariant(u32),
+    /// Type should be an variant/enum.
+    #[error(
+        "{0} type should be an variant/enum type. Make sure you are providing a valid metadata V14"
+    )]
+    InvalidType(String),
+}
+
+impl CodegenError {
+    /// Fetch the location for this error.
+    // Todo: Probably worth storing location outside of the variant,
+    // so that there's a common way to set a location for some error.
+    fn get_location(&self) -> Span {
+        match self {
+            Self::InvalidModule(span) => *span,
+            Self::TypeSubstitutionError(err) => err.get_location(),
+            Self::InvalidTypePath(_, err) => err.span(),
+            _ => proc_macro2::Span::call_site(),
+        }
+    }
+    /// Render the error as an invocation of syn::compile_error!.
+    pub fn into_compile_error(self) -> TokenStream2 {
+        let msg = self.to_string();
+        let span = self.get_location();
+        syn::Error::new(span, msg).into_compile_error()
+    }
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum FetchMetadataError {
+    #[error("Cannot decode hex value: {0}")]
+    DecodeError(#[from] hex::FromHexError),
+    #[error("Request error: {0}")]
+    RequestError(#[from] jsonrpsee::core::Error),
+    #[error("'{0}' not supported, supported URI schemes are http, https, ws or wss.")]
+    InvalidScheme(String),
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum TypeSubstitutionError {
+    /// Substitute "to" type must be an absolute path.
+    #[error("The 'to' type must be a path prefixed with 'crate::' or '::'")]
+    ExpectedAbsolutePath(Span),
+    /// Substitute types must have a valid path.
+    #[error("Substitute types must have a valid path.")]
+    EmptySubstitutePath(Span),
+    #[error("Expected the from/to type generics to have the form 'Foo<A,B,C..>'.")]
+    ExpectedAngleBracketGenerics(Span),
```

Comment:
nit: missing docs here and on the `FetchMetadataError`.

Reply:
Hmm good point; we need a missing docs lint on this crate :)